src/site_upload/check_if_complete/check_if_complete.py

```diff
@@ -0,0 +1,140 @@
+import datetime
+import json
+import logging
+import os
+from time import sleep
+
+import awswrangler
+import boto3
+import botocore
+
+from shared import enums, functions
+
+logger = logging.getLogger()
+logger.setLevel("INFO")
+s3_client = boto3.client("s3")
+sns_client = boto3.client("sns")
+g_client = boto3.client("glue")
+
+
+def check_if_complete(message) -> (bool, dict):
+    transaction = functions.get_s3_json_as_dict(
+        bucket=os.environ.get("BUCKET_NAME"),
+        key=(
+            f"{enums.BucketPath.META.value}/transactions/{message['site']}__{message['study']}.json"
+        ),
+        s3_client=s3_client,
+    )
+    upload_time = datetime.datetime.fromisoformat(transaction["uploaded_at"])
+    for filetype, source, suffix in [
+        ("cube", enums.BucketPath.AGGREGATE.value, "aggregate"),
+        ("annotated_cube", enums.BucketPath.AGGREGATE.value, "aggregate"),
+        ("flat", enums.BucketPath.FLAT.value, "flat"),
+    ]:
+        filenames = transaction.get(filetype, [])
+        for filename in filenames:
+            dp = filename.split("__")[1].split(".")[0]
+            if source == enums.BucketPath.FLAT.value:
+                key = (
+                    f"{source}/{message['study']}/{message['site']}/{message['study']}__{dp}__{message['site']}__{transaction['version']}/"
+                    f"{message['study']}__{dp}__{message['site']}__{suffix}.parquet"
+                )
+            else:
+                key = (
+                    f"{source}/{message['study']}/{message['study']}__{dp}/{message['study']}__{dp}__{transaction['version']}/"
+                    f"{message['study']}__{dp}__{suffix}.parquet"
+                )
+            try:
+                head = s3_client.head_object(Bucket=os.environ.get("BUCKET_NAME"), Key=key)
+            except s3_client.exceptions.ClientError:
+                return False, None
+            delta = head["LastModified"] - upload_time
+            # an average time to process from an upload to all the files being
+            # processed is a minute, and since this is largely parallel it shouldn't
+            # change by much; we'll set a large buffer just in case, which should
+            # still exclude dangling files from error states
+            if upload_time > head["LastModified"] or (delta.seconds >= 300 and delta.seconds < 0):
```

Comment:
1. You made the mistake everyone makes when dealing with timedelta objects - you want to look at `total_seconds()`, not `seconds` -- `timedelta(days=1).seconds == 0` but `timedelta(days=1).total_seconds() == 86400.0)`.
2. If upload time is newer, delta.total_seconds() is negative, yeah? So you don't need the first part of this check.
3. You have an `and` in the middle of the delta.seconds check that should never be true, yeah?
4. Python has a cool feature where you can do a triple comparison, which might simplify this logic: `if not (0 <= delta.total_seconds() < 300):`
5. This should have a test if it doesn't yet :smile: